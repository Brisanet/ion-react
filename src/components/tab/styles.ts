import stitches from '../../stitches.config';
import { spacing } from '../utils/spacing';

const { styled } = stitches;

export const TabStyles = styled('button', {
  fontStyle: 'normal',
  fontWeight: '600',
  color: '$neutral7',
  background: '$neutral1',
  cursor: 'pointer',
  display: 'flex',
  alignItems: 'center',
  gap: spacing(1),

  '&:hover': {
    color: '$primary5',
    background: '$primary1',
    borderColor: '$primary3',
  },
  '&:disabled': {
    color: '$neutral5',
    background: '$neutral3',
    borderColor: '$neutral5',
    cursor: 'not-allowed',
  },
  '&:focus': {
    color: '$primaryColor',
    background: '$neutral1',
    borderColor: '$primaryColor',
    fontStyle: 'normal',
    fontWeight: '600',
    cursor: 'pointer',
    display: 'flex',
    alignItems: 'center',
  },

  variants: {
    size: {
      sm: {
        padding: '8px 16px',
        fontSize: '12px',
        lineHeight: '16px',
        gap: spacing(1),
      },
      md: {
        padding: '10px 24px',
        fontSize: '14px',
        lineHeight: '20px',
        gap: spacing(1.5),
      },
      lg: {
        padding: '12px 32px',
        fontSize: '16px',
        lineHeight: '24px',
        gap: spacing(2),
      },
    },
    direction: {
      bottom: {
        borderBottom: '2px solid $neutral3',
        borderRadius: '6px 6px 0px 0px',
      },
      top: {
        borderTop: '2px solid $neutral3',
        borderRadius: '0px 0px 6px 6px',
      },
      right: {
        borderRight: '2px solid $neutral3',
      },
      left: {
        borderLeft: '2px solid $neutral3',
      },
    },
    selected: {
      true: {
        color: '$primaryColor',
      },
    },
  },

  compoundVariants: [
    {
      selected: true,
      css: {
        background: '$neutral1',
        borderColor: '$primaryColor',
        '&:hover': {
          color: '$primary5',
          background: '$primary1',
          borderColor: '$primary3',
          '&:active': {
            color: '$primary7',
            background: '$primary2',
            borderColor: '$primary5',
          },
          '&:disabled': {
            color: '$neutral5',
            background: '$neutral3',
            borderColor: '$neutral5',
            cursor: 'not-allowed',
          },
        },
      },
    },
  ],
});
