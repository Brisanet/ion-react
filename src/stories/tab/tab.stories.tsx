import { ComponentStory, ComponentMeta } from '@storybook/react';

import IonTab, { TabProps } from '../../components/tab/tab';

export default {
  title: 'Ion/Navigation/Tab',
  component: IonTab,
} as ComponentMeta<typeof IonTab>;

const Template: ComponentStory<typeof IonTab> = (args: TabProps) => (
  <IonTab {...args} />
);

export const TabBasic = Template.bind({});
TabBasic.args = {
  label: 'Custom label',
};
export const TabSmall = Template.bind({});
TabSmall.args = {
  label: 'Small',
  size: 'sm',
};

export const TabMedium = Template.bind({});
TabMedium.args = {
  label: 'Medium',
  size: 'md',
};

export const TabLarge = Template.bind({});
TabLarge.args = {
  label: 'Large',
  size: 'lg',
};

export const TabBottom = Template.bind({});
TabBottom.args = {
  label: 'Bottom',
  direction: 'bottom',
};

export const TabTop = Template.bind({});
TabTop.args = {
  label: 'Top',
  direction: 'top',
};

export const TabRight = Template.bind({});
TabRight.args = {
  label: 'Right',
  direction: 'right',
};

export const TabLeft = Template.bind({});
TabLeft.args = {
  label: 'Left',
  direction: 'left',
};

export const Disabled = Template.bind({});
Disabled.args = {
  label: 'Disabled',
  disabled: true,
};

export const Selected = Template.bind({});
Selected.args = {
  label: 'Selected',
  selected: true,
};

export const WithIcon = Template.bind({});
WithIcon.args = {
  label: 'Icon',
  icon: 'pencil',
};

export const WithBadge = Template.bind({});
WithBadge.args = {
  label: 'Badge',
  badge: { label: 10 },
};

export const IconAndBadge = Template.bind({});
IconAndBadge.args = {
  label: 'Badge',
  icon: 'pencil',
  badge: { label: 10 },
};
