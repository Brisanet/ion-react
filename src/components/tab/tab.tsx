import React, { useEffect, useState } from 'react';
import { SizeType } from '../../core/types/size';
import IonIcon from '../icons/icons';
import { iconType } from '../icons/svgs/icons';
import { TabStyles } from './styles';

type StitchesTabProps = React.ComponentProps<typeof TabStyles>;

type TabSizes = Exclude<SizeType, 'xs'>;

export type TabProps = {
  label: string;
  direction?: string;
  size?: TabSizes;
  selected?: boolean;
  disabled?: boolean;
  icon?: iconType;
  handleClick?: () => void;
} & StitchesTabProps;

const IonTab = ({
  label,
  icon,
  size = 'sm',
  direction = 'bottom',
  selected = false,
  disabled = false,
  handleClick,
}: TabProps) => {
  const [selectedState, setSelectedState] = useState<boolean>(selected);

  /**
   * If the @component IonTab is used individually outside
   * the @param handleClick function that changes the
   * 'selected' status will be defined to support individual IonTabs
   */
  if (handleClick === undefined) {
    handleClick = () => {
      setSelectedState(true);
    };
  }

  useEffect(() => {
    setSelectedState(selected);
  }, [selected]);

  const iconSize = {
    sm: 16,
    md: 20,
    lg: 24,
  };

  return (
    <>
      <TabStyles
        size={size}
        direction={direction}
        onClick={handleClick}
        selected={selectedState}
        disabled={disabled}
      >
        <div>
          {icon && <IonIcon type={icon} size={iconSize[`${size}`]} />}
          <span>{label}</span>
        </div>
      </TabStyles>
    </>
  );
};

export default IonTab;
