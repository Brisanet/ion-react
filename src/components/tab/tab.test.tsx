import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import IonTab from './tab';
import { TabProps } from './tab';

const clickEvent = jest.fn();
const defaultTab: TabProps = {
  label: 'Tab',
  handleClick: clickEvent,
};

function sut(props: TabProps = defaultTab): HTMLElement {
  render(<IonTab {...props} />);
  return screen.getByRole('button', { name: props.label });
}

describe('Tab', () => {
  describe('Default', () => {
    it('should render the Tab component', () => {
      const label = 'Historico';
      sut({ ...defaultTab, label: label });
      expect(screen.getByText(label)).toBeInTheDocument();
    });

    it('should execute user event when the Tab is clicked', async () => {
      await userEvent.click(sut({ ...defaultTab }));
      expect(clickEvent).toHaveBeenCalled();
    });

    it('should be contain the "selected-true" class when clicked without handleClick prop', async () => {
      const myTab = sut({ ...defaultTab, handleClick: undefined });
      await userEvent.click(myTab);
      expect(myTab.className).toContain('selected-true');
    });

    it('should be disabled', () => {
      expect(sut({ ...defaultTab, disabled: true })).toHaveAttribute(
        'disabled'
      );
    });

    it('should render tab without disabled by default', async () => {
      expect(sut({ ...defaultTab })).toBeEnabled();
    });

    it('should not call handeClick event when tab is disabled', async () => {
      await userEvent.click(sut({ ...defaultTab, disabled: true }));
      expect(clickEvent).toBeCalledTimes(0);
    });

    afterEach(() => {
      clickEvent.mockClear();
    });
  });

  describe('Custom Props', () => {
    const tabSizes: Array<TabProps['size']> = ['sm', 'md', 'lg'];
    it.each(tabSizes)('should render Tab with %s size', (size) => {
      const { className } = sut({ ...defaultTab, size: size });
      expect(className).toContain(`size-${size}`);
    });

    const tabDirections: Array<TabProps['direction']> = [
      'top',
      'left',
      'bottom',
      'right',
    ];
    it.each(tabDirections)('should render Tab with %s border', (direction) => {
      const { className } = sut({ ...defaultTab, direction: direction });
      expect(className).toContain(`direction-${direction}`);
    });
  });

  describe('With Icon', () => {
    const iconName = 'pencil';

    it('should render pencil icon', () => {
      sut({ ...defaultTab, icon: iconName });
      expect(screen.getByTestId(`ion-icon-${iconName}`)).toBeInTheDocument();
    });

    it('should render sm icon size by default', () => {
      const sm = '16';
      sut({ ...defaultTab, icon: iconName });
      const Tab = screen.getByTestId(`ion-icon-${iconName}`);
      expect(Tab).toHaveAttribute('height', sm);
    });

    it('should render icon md when tab is md', () => {
      const mdSize = '20';
      const mdIconName = 'alert';
      sut({ ...defaultTab, name: 'tab-icon', size: 'md', icon: mdIconName });
      expect(screen.getByTestId(`ion-icon-${mdIconName}`)).toHaveAttribute(
        'height',
        mdSize
      );
    });
  });
});
