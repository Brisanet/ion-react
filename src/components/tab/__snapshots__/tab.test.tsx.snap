// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab Custom Props should render Tab with bottom border 1`] = `
.c0 {
  font-weight: 600;
  color: #505566;
  background: #fcfcfd;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 8px;
  padding: 8px 16px;
  font-size: 1.2rem;
  line-height: 1.6rem;
  border-bottom: 2px solid;
  border-radius: 6px 6px 0px 0px;
  border-color: #ced2db;
}

.c0:focus-visible {
  outline: 2px solid #146ff5;
  outline-offset: 2px;
}

.c0 svg {
  fill: #505566;
}

.c0:hover,
.c0:focus-visible {
  color: #146ff5;
  background: #ebf3fe;
  border-color: #84b4fa;
}

.c0:hover svg,
.c0:focus-visible svg {
  fill: #146ff5;
}

.c0:active {
  color: #06439d;
  background: #b5d2fc;
  border-color: #146ff5;
}

.c0:active svg {
  fill: #06439d;
}

.c0:disabled {
  color: #aeb2bd;
  background: #e4e6eb;
  border-color: #aeb2bd;
  cursor: not-allowed;
}

.c0:disabled svg {
  fill: #aeb2bd;
}

<div>
  <button
    class="c0"
    data-testid="ion-tab"
  >
    <span>
      Tab
    </span>
  </button>
</div>
`;

exports[`Tab Custom Props should render Tab with left border 1`] = `
.c0 {
  font-weight: 600;
  color: #505566;
  background: #fcfcfd;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 8px;
  padding: 8px 16px;
  font-size: 1.2rem;
  line-height: 1.6rem;
  border-left: 2px solid;
  border-radius: 0px 6px 6px 0px;
  border-color: #ced2db;
}

.c0:focus-visible {
  outline: 2px solid #146ff5;
  outline-offset: 2px;
}

.c0 svg {
  fill: #505566;
}

.c0:hover,
.c0:focus-visible {
  color: #146ff5;
  background: #ebf3fe;
  border-color: #84b4fa;
}

.c0:hover svg,
.c0:focus-visible svg {
  fill: #146ff5;
}

.c0:active {
  color: #06439d;
  background: #b5d2fc;
  border-color: #146ff5;
}

.c0:active svg {
  fill: #06439d;
}

.c0:disabled {
  color: #aeb2bd;
  background: #e4e6eb;
  border-color: #aeb2bd;
  cursor: not-allowed;
}

.c0:disabled svg {
  fill: #aeb2bd;
}

<div>
  <button
    class="c0"
    data-testid="ion-tab"
  >
    <span>
      Tab
    </span>
  </button>
</div>
`;

exports[`Tab Custom Props should render Tab with lg size 1`] = `
.c0 {
  font-weight: 600;
  color: #505566;
  background: #fcfcfd;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 8px;
  padding: 12px 32px;
  font-size: 1.6rem;
  line-height: 2.4rem;
  border-bottom: 2px solid;
  border-radius: 6px 6px 0px 0px;
  border-color: #ced2db;
}

.c0:focus-visible {
  outline: 2px solid #146ff5;
  outline-offset: 2px;
}

.c0 svg {
  fill: #505566;
}

.c0:hover,
.c0:focus-visible {
  color: #146ff5;
  background: #ebf3fe;
  border-color: #84b4fa;
}

.c0:hover svg,
.c0:focus-visible svg {
  fill: #146ff5;
}

.c0:active {
  color: #06439d;
  background: #b5d2fc;
  border-color: #146ff5;
}

.c0:active svg {
  fill: #06439d;
}

.c0:disabled {
  color: #aeb2bd;
  background: #e4e6eb;
  border-color: #aeb2bd;
  cursor: not-allowed;
}

.c0:disabled svg {
  fill: #aeb2bd;
}

<div>
  <button
    class="c0"
    data-testid="ion-tab"
  >
    <span>
      Tab
    </span>
  </button>
</div>
`;

exports[`Tab Custom Props should render Tab with md size 1`] = `
.c0 {
  font-weight: 600;
  color: #505566;
  background: #fcfcfd;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 8px;
  padding: 10px 24px;
  font-size: 1.4rem;
  line-height: 2rem;
  border-bottom: 2px solid;
  border-radius: 6px 6px 0px 0px;
  border-color: #ced2db;
}

.c0:focus-visible {
  outline: 2px solid #146ff5;
  outline-offset: 2px;
}

.c0 svg {
  fill: #505566;
}

.c0:hover,
.c0:focus-visible {
  color: #146ff5;
  background: #ebf3fe;
  border-color: #84b4fa;
}

.c0:hover svg,
.c0:focus-visible svg {
  fill: #146ff5;
}

.c0:active {
  color: #06439d;
  background: #b5d2fc;
  border-color: #146ff5;
}

.c0:active svg {
  fill: #06439d;
}

.c0:disabled {
  color: #aeb2bd;
  background: #e4e6eb;
  border-color: #aeb2bd;
  cursor: not-allowed;
}

.c0:disabled svg {
  fill: #aeb2bd;
}

<div>
  <button
    class="c0"
    data-testid="ion-tab"
  >
    <span>
      Tab
    </span>
  </button>
</div>
`;

exports[`Tab Custom Props should render Tab with right border 1`] = `
.c0 {
  font-weight: 600;
  color: #505566;
  background: #fcfcfd;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 8px;
  padding: 8px 16px;
  font-size: 1.2rem;
  line-height: 1.6rem;
  border-right: 2px solid;
  border-radius: 6px 0px 0px 6px;
  border-color: #ced2db;
}

.c0:focus-visible {
  outline: 2px solid #146ff5;
  outline-offset: 2px;
}

.c0 svg {
  fill: #505566;
}

.c0:hover,
.c0:focus-visible {
  color: #146ff5;
  background: #ebf3fe;
  border-color: #84b4fa;
}

.c0:hover svg,
.c0:focus-visible svg {
  fill: #146ff5;
}

.c0:active {
  color: #06439d;
  background: #b5d2fc;
  border-color: #146ff5;
}

.c0:active svg {
  fill: #06439d;
}

.c0:disabled {
  color: #aeb2bd;
  background: #e4e6eb;
  border-color: #aeb2bd;
  cursor: not-allowed;
}

.c0:disabled svg {
  fill: #aeb2bd;
}

<div>
  <button
    class="c0"
    data-testid="ion-tab"
  >
    <span>
      Tab
    </span>
  </button>
</div>
`;

exports[`Tab Custom Props should render Tab with sm size 1`] = `
.c0 {
  font-weight: 600;
  color: #505566;
  background: #fcfcfd;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 8px;
  padding: 8px 16px;
  font-size: 1.2rem;
  line-height: 1.6rem;
  border-bottom: 2px solid;
  border-radius: 6px 6px 0px 0px;
  border-color: #ced2db;
}

.c0:focus-visible {
  outline: 2px solid #146ff5;
  outline-offset: 2px;
}

.c0 svg {
  fill: #505566;
}

.c0:hover,
.c0:focus-visible {
  color: #146ff5;
  background: #ebf3fe;
  border-color: #84b4fa;
}

.c0:hover svg,
.c0:focus-visible svg {
  fill: #146ff5;
}

.c0:active {
  color: #06439d;
  background: #b5d2fc;
  border-color: #146ff5;
}

.c0:active svg {
  fill: #06439d;
}

.c0:disabled {
  color: #aeb2bd;
  background: #e4e6eb;
  border-color: #aeb2bd;
  cursor: not-allowed;
}

.c0:disabled svg {
  fill: #aeb2bd;
}

<div>
  <button
    class="c0"
    data-testid="ion-tab"
  >
    <span>
      Tab
    </span>
  </button>
</div>
`;

exports[`Tab Custom Props should render Tab with top border 1`] = `
.c0 {
  font-weight: 600;
  color: #505566;
  background: #fcfcfd;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 8px;
  padding: 8px 16px;
  font-size: 1.2rem;
  line-height: 1.6rem;
  border-top: 2px solid;
  border-radius: 0px 0px 6px 6px;
  border-color: #ced2db;
}

.c0:focus-visible {
  outline: 2px solid #146ff5;
  outline-offset: 2px;
}

.c0 svg {
  fill: #505566;
}

.c0:hover,
.c0:focus-visible {
  color: #146ff5;
  background: #ebf3fe;
  border-color: #84b4fa;
}

.c0:hover svg,
.c0:focus-visible svg {
  fill: #146ff5;
}

.c0:active {
  color: #06439d;
  background: #b5d2fc;
  border-color: #146ff5;
}

.c0:active svg {
  fill: #06439d;
}

.c0:disabled {
  color: #aeb2bd;
  background: #e4e6eb;
  border-color: #aeb2bd;
  cursor: not-allowed;
}

.c0:disabled svg {
  fill: #aeb2bd;
}

<div>
  <button
    class="c0"
    data-testid="ion-tab"
  >
    <span>
      Tab
    </span>
  </button>
</div>
`;

exports[`Tab Custom Props should render selected tab 1`] = `
.c0 {
  font-weight: 600;
  color: #505566;
  background: #fcfcfd;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 8px;
  padding: 8px 16px;
  font-size: 1.2rem;
  line-height: 1.6rem;
  border-bottom: 2px solid;
  border-radius: 6px 6px 0px 0px;
  border-color: #ced2db;
  color: #0858ce;
  border-color: #0858ce;
}

.c0:focus-visible {
  outline: 2px solid #146ff5;
  outline-offset: 2px;
}

.c0 svg {
  fill: #505566;
}

.c0:hover,
.c0:focus-visible {
  color: #146ff5;
  background: #ebf3fe;
  border-color: #84b4fa;
}

.c0:hover svg,
.c0:focus-visible svg {
  fill: #146ff5;
}

.c0:active {
  color: #06439d;
  background: #b5d2fc;
  border-color: #146ff5;
}

.c0:active svg {
  fill: #06439d;
}

.c0:disabled {
  color: #aeb2bd;
  background: #e4e6eb;
  border-color: #aeb2bd;
  cursor: not-allowed;
}

.c0:disabled svg {
  fill: #aeb2bd;
}

.c0 svg {
  fill: #0858ce;
}

<div>
  <button
    class="c0"
    data-testid="ion-tab"
  >
    <span>
      Tab
    </span>
  </button>
</div>
`;
